### PR TITLE
add aggregate expression check

### PR DESCRIPTION
**Pull request by jyizheng**
## Changelog

- Bug Fix

## Related Issues

#404

### CODE REVIEW

**Comment by BohuTANG** [comment]
Hi Andy,

For the query if the projection(here `a`) in group-by expression, it's also valid .

The query run in ClickHouse:
```
select number%3 as a, min(number), max(number) from numbers_mt(1000) where number>3 group by a order by a;

SELECT
    number % 3 AS a,
    min(number),
    max(number)
FROM numbers_mt(1000)
WHERE number > 3
GROUP BY a
ORDER BY a ASC

Query id: 2a581bc7-0241-44e7-8f38-8ca1cd4f528a

┌─a─┬─min(number)─┬─max(number)─┐
│ 0 │           6 │         999 │
│ 1 │           4 │         997 │
│ 2 │           5 │         998 │
└───┴─────────────┴─────────────┘

3 rows in set. Elapsed: 0.004 sec. Processed 1.00 thousand rows, 8.00 KB (229.01 thousand rows/s., 1.83 MB/s.) 

```

**Comment by jyizheng** [comment]
This PR is ready. Please take a loot at it.

File: common/planners/src/plan_builder.rs [hunk 1]
```diff
@@ -106,6 +130,29 @@ impl PlanBuilder {
         let input_schema = self.plan.schema();
         let aggr_projection_fields = self.exprs_to_fields(&aggr_expr, &input_schema)?;
 
+        // Aggregator check.
+        for e_aggr in &aggr_expr {
```
Comment (BohuTANG) [inline common/planners/src/plan_builder.rs:134]: Good.
But here we need a plan_builder uint test to against the check.
From the code, needs:
1. the passed case
2. the failed case

Comment (jyizheng) [reply common/planners/src/plan_builder.rs:134]: Sure. No problem, happy to work on it.

Comment (jyizheng) [reply common/planners/src/plan_builder.rs:134]: I added three tests in plan_aggregator_test.rs. Do I miss anything?

**Comment by BohuTANG** [comment]
Hi @jyizheng ,

This merge conflict is a bit weird, how do you do it?
If our barnch is confict with the master, the normal way:
1. git checkout master
2. git pull -ap
3. git checkout your-branch
4. git merge master
5. handle the conflict follow https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line

So, if we git commit the resolving confict, we will take the master commit to your branch, not the diff files.

**Comment by jyizheng** [comment]
> Hi @jyizheng ,
> 
> This merge conflict is a bit weird, how do you do it?
> If our barnch is confict with the master, the normal way:
> 
> 1. git checkout master
> 2. git pull -ap
> 3. git checkout your-branch
> 4. git merge master
> 5. handle the conflict follow https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line
> 
> So, if we git commit the resolving confict, we will take the master commit to your branch, not the diff files.
> 
> The final will like(rebase is based on your `fix cargo fmt`):
> https://github.com/datafuselabs/datafuse/pull/452/files

Thanks for the guidance.  My steps are similar to what you have shown above.
I merged the master into my branch and manually fix any conflicts.

I just checked the commit "fix merge conflict". Yes, it is so weird.
I am not sure if I missed anything step during the merge.

**Comment by BohuTANG** [comment]
> > Hi @jyizheng ,
> > This merge conflict is a bit weird, how do you do it?
> > If our barnch is confict with the master, the normal way:
> > 
> > 1. git checkout master
> > 2. git pull -ap
> > 3. git checkout your-branch
> > 4. git merge master
> > 5. handle the conflict follow https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line
> > 
> > So, if we git commit the resolving confict, we will take the master commit to your branch, not the diff files.
> > The final will like(rebase is based on your `fix cargo fmt`):
> > https://github.com/datafuselabs/datafuse/pull/452/files
> 
> Thanks for the guidance. My steps are similar to what you have shown above.
> I merged the master into my branch and manually fix any conflicts.
> 
> I just checked the commit "fix merge conflict". Yes, it is so weird.
> I am not sure if I missed anything step during the merge.

Hello Andy,

The reason is your rebase you new commits to the master commit.
So git willl create a new commit and take the diff files.
Found that you have fixed that, cool.

**Comment by BohuTANG** [comment]
BTW, to fix the stateless failed, you should merge the master again(master fixed the issue).

**Comment by jyizheng** [comment]
> BTW, to fix the stateless failed, you should merge the master again(master fixed the issue).

"SELECT number%3 as c1 FROM numbers_mt(10) where number > 2 group by number order by c1;"
The SQL above is not handled properly right now in this PR. I am trying to figure out how to let the aggregate() function
know an expression is used by `order by`.

**Comment by jyizheng** [comment]
> > BTW, to fix the stateless failed, you should merge the master again(master fixed the issue).
> 
> "SELECT number%3 as c1 FROM numbers_mt(10) where number > 2 group by number order by c1;"
> The SQL above is not handled properly right now in this PR. I am trying to figure out how to let the aggregate() function
> know an expression is used by `order by`.

Is it a good idea to build a `order by` plan first here:

https://github.com/datafuselabs/datafuse/blob/3e9da2ee47913511141130b9358645129af96640/fusequery/query/src/sql/plan_parser.rs#L229
, and then pass the order-by plan to aggregate so that aggregate() function in plan_builder.rs can know if an aggregate expression shows up in `order by` ?

**Comment by BohuTANG** [comment]
> > > BTW, to fix the stateless failed, you should merge the master again(master fixed the issue).
> > 
> > 
> > "SELECT number%3 as c1 FROM numbers_mt(10) where number > 2 group by number order by c1;"
> > The SQL above is not handled properly right now in this PR. I am trying to figure out how to let the aggregate() function
> > know an expression is used by `order by`.
> 
> Is it a good idea to build a `order by` plan first here:
> 
> https://github.com/datafuselabs/datafuse/blob/3e9da2ee47913511141130b9358645129af96640/fusequery/query/src/sql/plan_parser.rs#L229
> 
> 
> , and then pass the order-by plan to aggregate so that aggregate() function in plan_builder.rs can know if an aggregate expression shows up in `order by` ?

It looks that the `order by` plan expression need to be check in optimizer.
So what's the problem when we check it?

**Comment by BohuTANG** [comment]
https://github.com/datafuselabs/datafuse/issues/460

It looks like we need to abstract out an optimizer rule, such as alias push down, for filter/groupby/orderby all.
Currently, all alias push down logic is mixed in each rules.

**Comment by BohuTANG** [comment]
Hi @jyizheng 

There is a refactor PR(for code clean) merged to master:
https://github.com/datafuselabs/datafuse/pull/462

File: common/planners/src/plan_builder.rs [hunk 2]
```diff
@@ -107,6 +131,29 @@ impl PlanBuilder {
...
+                let mut in_group_by = false;
+                // Check if e_aggr is in group-by's list
+                for e_group in &group_expr {
+                    if aggr_group_expr_eq(&e_aggr, &e_group, &input_schema)? {
+                        in_group_by = true;
+                        break;
+                    }
+                }
+                if !in_group_by {
+                    return Result::Err(ErrorCodes::UnknownException(format!(
```
Comment (zhang2014) [inline common/planners/src/plan_builder.rs:149]: We need a new error code type.

**Comment by BohuTANG** [comment]
Hi @jyizheng 

No need rebase your commit to one of master commit(it will bring master changes to your branch), just keep the merge commit is ok.

**Comment by jyizheng** [comment]
> Hi @jyizheng
> 
> No need rebase your commit to one of master commit(it will bring master changes to your branch), just keep the merge commit is ok.

Below are my steps:

$git checkout upstream/master
$git pull upstream master
$git checkout master
$git merge upstream/master
$git push origin master
$git checkout _this_branch_
$git merge master
$git push origin _this_branch_

Which step I should fix?

**Comment by BohuTANG** [comment]
The `file changes` is OK now.

**Comment by jyizheng** [comment]
> The `file changes` is OK now.

I see. Probably because of the github-action after the commit.

File: common/planners/src/plan_builder.rs (hunk 2 above)
Comment (jyizheng) [reply common/planners/src/plan_builder.rs]: fixed

**Comment by jyizheng** [comment]
@BohuTANG @zhang2014  Could you please review this PR?

File: common/planners/src/plan_aggregator_test.rs [hunk 3]
```diff
@@ -2,29 +2,107 @@
...
-    });
-    let expect = "Projection: sumx:UInt64\
-    \n  AggregatorFinal: groupBy=[[]], aggr=[[sum([number]) as sumx]]\
-    \n    AggregatorPartial: groupBy=[[]], aggr=[[sum([number]) as sumx]]\
-    \n      ReadDataSource: scan partitions: [8], scan schema: [number:UInt64], statistics: [read_rows: 10000, read_bytes: 80000]";
-    let actual = format!("{:?}", explain);
-    assert_eq!(expect, actual);
-    Ok(())
+#[cfg(test)]
+mod tests {
```
Comment (BohuTANG) [inline common/planners/src/plan_aggregator_test.rs:6]: Using this mode will be more compact for multi case tests:
https://github.com/datafuselabs/datafuse/blob/c64351c804ce1813f3a952faa2d203028ac57d23/common/planners/src/plan_builder_test.rs#L12-L41

File: common/planners/src/plan_builder.rs [hunk 4]
```diff
@@ -38,6 +40,40 @@ pub struct PlanBuilder {
...
+    match aggr {
+        ExpressionPlan::Alias(alias, plan) => {
+            if group_by_names.contains(alias) {
+                return Ok(true);
+            } else {
+                return aggr_group_expr_eq(plan, group_by_names, input_schema);
+            }
+        }
+        _ => {
+            let aggr_str = String::from(format!("{:?}", aggr).as_str());
```
Comment (BohuTANG) [inline common/planners/src/plan_builder.rs:60]: let aggr_str = format!("{:?}", aggr);

File: common/planners/src/plan_builder.rs [hunk 5]
```diff
@@ -108,6 +156,25 @@ impl PlanBuilder {
         let input_schema = self.plan.schema();
         let aggr_projection_fields = self.exprs_to_fields(&aggr_expr, &input_schema)?;
 
+        // Aggregator check.
+        let mut group_by_names = HashSet::new();
```
[the remark this answers is not in the document]
Comment (jyizheng) [reply common/planners/src/plan_builder.rs:195]: Yes, it is implemented in `aggregate`.

File: common/planners/src/plan_aggregator_test.rs (hunk 3 above)
Comment (jyizheng) [reply common/planners/src/plan_aggregator_test.rs:6]: fixed

File: common/planners/src/plan_builder.rs (hunk 4 above)
Comment (jyizheng) [reply common/planners/src/plan_builder.rs]: Fixed

File: common/planners/src/plan_rewriter.rs [hunk 6]
```diff
@@ -212,6 +213,31 @@ impl PlanRewriter {
...
+    /// Get the leaves of an expression.
+    pub fn expression_plan_columns(expr: &ExpressionPlan) -> Result<Vec<ExpressionPlan>> {
+        Ok(match expr {
+            ExpressionPlan::Alias(_, expr) => Self::expression_plan_columns(expr)?,
+            ExpressionPlan::Column(_) => vec![expr.clone()],
+            ExpressionPlan::Literal(_) => vec![],
+            ExpressionPlan::BinaryExpression { left, right, .. } => {
+                let l = Self::expression_plan_columns(left)?;
+                let r = Self::expression_plan_columns(right)?;
+                concat(vec![l, r])
```
[the remark this answers is not in the document]
Comment (jyizheng) [reply common/planners/src/plan_rewriter.rs]: Fixed. I am using `append`.

**Review by BohuTANG — APPROVED** [review]
LGTM
Thank you @jyizheng

**Pull request merged**